Add or condition support.

yoshisatoyanagisawa opened the pull request:
The ServiceWorker static routing API has a or conditional syntax. This is a specification update to support that.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/7.html" title="Last updated on Jan 25, 2024, 7:06 AM UTC (d331f41)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/7/9c57f5f...d331f41.html" title="Last updated on Jan 25, 2024, 7:06 AM UTC (d331f41)">Diff</a>

yoshisatoyanagisawa committed:
Add or condition support.

yoshisatoyanagisawa commented:
@domenic @sisidovski Can I ask you to review the change?
Thank you in advance.

sisidovski commented on docs/index.bs:
```diff
@@ -1567,6 +1567,8 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
         RequestDestination requestDestination;
         RunningStatus runningStatus;
+
+        sequence&lt;RouterCondition&gt; _or;
```
I'm unsure this has to have a leading underscore. https://webidl.spec.whatwg.org/#idl-names

This underscore in the chromium implementation actually comes from the chromium Web IDL compiler, but Web IDL spec itself doesn't think `or` is a reserved keyword?

domenic replied:
I think it is reserved: https://webidl.spec.whatwg.org/#index-prod-UnionType has it in monospaced font, which I think means it's a terminal symbol and thus the tokenizer will pick it up as part of the grammar? I'm pretty rusty on my grammar stuff though.

But we have had to do similar escaping in the past, e.g. note the definition of `_any` in https://dom.spec.whatwg.org/#interface-AbortSignal . So this makes sense to me.

sisidovski replied:
I couldn't find the whole list. That makes sense, thanks!

domenic commented on docs/index.bs:
```diff
@@ -3237,30 +3239,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      1. If |condition|["{{RouterCondition/_or}}"] [=map/exists=], then:
+          1. If |hasCondition| is true, return false.
+
+              Note: For ease of understanding the router rule, the "or" condition is mutual exclusive with other conditions.
```
```suggestion
              Note: For ease of understanding the router rule, the "or" condition is mutually exclusive with other conditions.
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3237,30 +3239,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+              Note: For ease of understanding the router rule, the "or" condition is mutual exclusive with other conditions.
+
+          1. Let |or conditions| be |condition|["{{RouterCondition/_or}}"].
```
Nit: `|orConditions|` / `|orCondition|` here and below is the more modern style.

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3237,30 +3239,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      Note: if there are multiple conditions, all conditions will be matched to return the source.
```
This note probably makes more sense inside the loop in Get Router Source, instead of here in the algorithm that's focused on only a single condition?

yoshisatoyanagisawa replied:
This explanation has been added as an answer to https://github.com/yoshisatoyanagisawa/ServiceWorker/pull/5#discussion_r1453049376.

People can write something like:
```
condition: {
  urlPattern: "/test",
  runningStatus: "running"
}
```

This function will return true if both `urlPattern` and `runningStatus` are fulfilled.  I tried to explain that here.

yoshisatoyanagisawa replied:
But return "source" sounds wrong here X(

yoshisatoyanagisawa replied:
I have updated the explanation.

domenic commented on docs/index.bs:
```diff
@@ -3237,30 +3239,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. If |runningStatus| is {{RunningStatus/"not-running"}}, and |serviceWorker| is [=running=], return false.
+      1. If |condition|["{{RouterCondition/_or}}"] [=map/exists=], then:
+          1. Let |or conditions| be |condition|["{{RouterCondition/_or}}"].
+          1. Let |or condition matched| be false.
```
Nit: you can simplify this as follows:

> 1. For each orCondition of orConditions:
>    1. If running ... returns true, then return true.
> 1. Return false.

yoshisatoyanagisawa replied:
Ah, it reminds me https://github.com/WICG/service-worker-static-routing-api/issues/9
`or: []` always returns false.
I have applied the change.

domenic reviewed (approved):
LGTM with nits

yoshisatoyanagisawa committed:
Reflect comments.

- mutual to mutually.
- |or condition| or |or conditions| to |orCondition| or |orConditions|.
- simplify the for loop upon WICG/service-worker-static-routing-api#9.

yoshisatoyanagisawa committed:
Revised note to make it match with the moved location.

yoshisatoyanagisawa commented:
Thanks for the review.  I am going to merge this.
